packages/alfa-rules/src/common/diagnostic/with-bounding-box.ts

```diff
@@ -12,35 +18,92 @@ export class WithBoundingBox extends WithName {
     message: string,
     name: string,
     box: Rectangle,
+    condition: Either<
+      WithBoundingBox.UACondition,
+      WithBoundingBox.SizeAndSpacingCondition
+    >,
+    tooCloseNeighbors: Iterable<Element>,
   ): WithBoundingBox;
 
   public static of(
     message: string,
     name?: string,
     box?: Rectangle,
+    condition?: Either<
+      WithBoundingBox.UACondition,
+      WithBoundingBox.SizeAndSpacingCondition
+    >,
+    tooCloseNeighbors?: Iterable<Element>,
   ): Diagnostic {
     if (name === undefined) {
       return new Diagnostic(message);
     }
 
-    if (box === undefined) {
+    if (
+      box === undefined ||
+      condition === undefined ||
+      tooCloseNeighbors === undefined
+    ) {
       return new WithName(message, name);
     }
 
-    return new WithBoundingBox(message, name, box);
+    return new WithBoundingBox(
+      message,
+      name,
+      box,
+      condition,
+      tooCloseNeighbors,
+    );
   }
 
   protected readonly _box: Rectangle;
+  protected readonly _condition: Either<
+    WithBoundingBox.UACondition,
+    WithBoundingBox.SizeAndSpacingCondition
+  >;
+  protected readonly _tooCloseNeighbors: Sequence<Element>;
 
-  protected constructor(message: string, name: string, box: Rectangle) {
+  protected constructor(
+    message: string,
+    name: string,
+    box: Rectangle,
+    condition: Either<
+      WithBoundingBox.UACondition,
+      WithBoundingBox.SizeAndSpacingCondition
+    >,
+    tooCloseNeighbors: Iterable<Element>,
+  ) {
     super(message, name);
     this._box = box;
+
+    // Copy the objects to ensure they are immutable from the outside.
+    this._condition = condition.either(
+      (uaCond) => Either.left({ ua: uaCond.ua }),
+      (sizeAndSpacingCond) =>
+        Either.right({
+          size: sizeAndSpacingCond.size,
+          spacing: sizeAndSpacingCond.spacing,
+        }),
+    );
+
+    this._tooCloseNeighbors = Sequence.from(tooCloseNeighbors);
   }
 
   public get box(): Rectangle {
     return this._box;
   }
 
+  public get condition(): Either<
+    WithBoundingBox.UACondition,
+    WithBoundingBox.SizeAndSpacingCondition
+  > {
+    return this._condition;
+  }
+
+  public get tooCloseNeighbors(): Sequence<Element> {
```

Comment:
```suggestion
  public get tooCloseNeighbors(): Iterable<Element> {
```
We usually expose stuff as `Iterable` to avoid exposing implementation details (i.e.  we can switch to another kind of iterable if need be without breaking consumers).